Content.Shared/Trigger/Systems/ScramOnTriggerSystem.cs

```diff
@@ -60,81 +59,52 @@ protected override void OnTrigger(Entity<ScramOnTriggerComponent> ent, EntityUid
+            // then we multiply that with a randomly angled vector.
+            // The distance is then reduced percentage-wise based on the current try count,
+            // so subsequent tries are closer and closer towards the entity. This is
+            // beneficial for smaller maps, especially when the radius is large.
+            var distance = radius * MathF.Sqrt(_random.NextFloat()) * (1 - (float)i / tries);
+            var tempTargetCoords = userCoords.Offset(_random.NextAngle().ToVec() * distance);
+
+            // Skip if there's no grid at the target coordinates
+            if (!_mapManager.TryFindGridAt(_transform.ToMapCoordinates(tempTargetCoords), out var gridUid, out var grid))
+                continue;
+
+            // Check every anchored entity on the tile for something that we don't want to
+            // teleport into, and skip if we find something.
+            var tileValid = true;
+            foreach (var entity in _map.GetAnchoredEntities(gridUid, grid, tempTargetCoords))
             {
-                tileList.Add(tile.GridIndices);
-            }
+                if (!_physicsQuery.TryGetComponent(entity, out var body))
+                    continue;
 
-            while (tileList.Count != 0)
-            {
-                var tile = tileList.RemoveSwap(_random.Next(tileList.Count));
-                valid = true;
-                foreach (var entity in _map.GetAnchoredEntities(targetGrid.Value.Owner, targetGrid.Value.Comp,
-                             tile))
-                {
-                    if (!_physicsQuery.TryGetComponent(entity, out var body))
-                        continue;
-
-                    if (body.BodyType != BodyType.Static ||
-                        !body.Hard ||
-                        (body.CollisionLayer & (int)CollisionGroup.MobMask) == 0)
-                        continue;
-
-                    valid = false;
-                    break;
-                }
-
-                if (valid)
-                {
-                    targetCoords = new EntityCoordinates(targetGrid.Value.Owner,
-                        _map.TileCenterToVector(targetGrid.Value, tile));
-                    break;
-                }
-            }
+                if (body.BodyType != BodyType.Static ||
+                    !body.Hard ||
+                    (body.CollisionLayer & (int)CollisionGroup.MobMask) == 0)
+                    continue;
 
-            if (valid || _targetGrids.Count == 0) // if we don't do the check here then PickAndTake will blow up on an empty set.
+                tileValid = false;
                 break;
+            }
 
-            targetGrid = _random.GetRandom().PickAndTake(_targetGrids);
-        } while (true);
+            if (!tileValid)
+                continue;
+
+            targetCoords = tempTargetCoords;
+            break;
+        }
 
         return targetCoords;
```

Comment:
This can all be simplified with `IsTileBlocked`.

```suggestion
            if (!_turfSystem.TryGetTileRef(tempTargetCoords, out var tileRef) || _turfSystem.IsTileBlocked(tileRef, CollisionGroup.MobMask))
              continue;
            
            targetCoords = tempTargetCoords;
            break;
          }
             
          return targetCoords;
```

Reply:
this should just get the collision mask of the entity being TP'd instead of using MobMask unconditionally